### ISSUE TITLE
fix ave/atom etc should produce standard deviation

### ISSUE THREAD

**Issue by andeplane** [lammps/lammps#251]
When using the ave/atom (and similar averaging commands) it makes sense to be able to produce the standard deviation for the quantity. 

For instance `stddev yes` will add N extra columns with the standard deviations for each N value.

What is your view on this?

**Comment by rjones-sandia** [comment]
I think variances would be super useful given the stat mech context. I wrote a version of fix ave/time to use an online/real time estimate of the variance. I'd would like to add that to lammps proper

Reese Jones
rjones@sandia.gov<mailto:rjones@sandia.gov> 925 294-4744
"What's gone and what's past help should be past grief" - Paulina, The Winter's Tale.

On Nov 3, 2016, at 12:58 PM, Anders Hafreager <notifications@github.com<mailto:notifications@github.com>> wrote:


When using the ave/atom (and similar averaging commands) it makes sense to be able to produce the standard deviation for the quantity.

For instance stddev yes will add N extra columns with the standard deviations for each N value.

What is your view on this?

-

**Comment by athomps** [comment]
I agree. Providing this as an option to fix ave/time would be very beneficial.  I never again want to figure out the correct form of this expression:

varx = (sumxsq - sumx*sumx/n)/n

Aidan



--
      Aidan P. Thompson
      01444 Multiscale Science
      Sandia National Laboratories
      PO Box 5800, MS 1322      Phone: 505-844-9702
      Albuquerque, NM 87185     Fax  : 505-845-7442
      E-mail:athomps@sandia.gov Cell : 505-218-1011

**Comment by andeplane** [comment]
Ok, I can start thinking about implementing this. I remember a technique where you don't get troubles when adding small quantities to a large cumulative sum.

**Comment by rjones-sandia** [comment]
the online/realtime algorithm is s bit more complex. give me a minute and I’ll try to dig up the code and a bit of documentation

Reese (Jones) 925 294-4744 rjones@sandia.gov<mailto:rjones@sandia.gov>






On Nov 3, 2016, at 1:19 PM, Anders Hafreager <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok, I can start thinking about implementing this. I remember a technique where you don't get troubles when adding small quantities to a large cumulative sum.

**Comment by rjones-sandia** [comment]
ok
  I think the algorithm embedded in the attached code (spawned from a now ancient version of fix_ave_time) is basically the one described in :

https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm

I used this extensively to get error estimates of time averages — i.e. I think it works fine.
Reese

**Comment by andeplane** [comment]
Ahh perfect! So am I understanding correctly if you have an old implementation in fix_ave_time which is not in the current version?

**Comment by andeplane** [comment]
@rjones-sandia could you send me the code you mentioned directly on email? Github ignores attachments when replying to issues through emails. My email is andershaf@gmail.com.

**Comment by andeplane** [comment]
No updates. I haven't gotten any reply from @rjones-sandia yet and I don't want to reimplement it since most of it is already done.

**Comment by andeplane** [comment]
@rjones-sandia ping?

**Comment by andeplane** [comment]
@sjplimp @athomps could you ask @rjones-sandia about this?

**Comment by zqex** [comment]
This comes with some caveats. While its easy to calculate some number < (x-<x>)^2> for some series of data, this is only the variance, IFF the sampled x'es are statistically independent, which depends on details of the observable and the system.

To estimate the variance, data should be blocked and the convergence after subsequent blocking transformations should be studied. See e.g. the algorithm in Flyvbjerg Pedersen J.Chem.Phys. 1, 461 (1989).  http://dx.doi.org/10.1063/1.457480  Not that trivial...

**Comment by Bibobu** [comment]
Part of this feature should be covered by #4527.

**Comment by alphataubio** [comment]
@akohlmey you can close this since `fix ave/moments` from @Bibobu implements this functionality
